### PR TITLE
Add tiered ticket types

### PR DESCRIPTION
**Pull request by mikethetike**
### Migrations
 * 20190401121212_add_parent_ticket_type_id

### Environment Variables
 * No change

### CODE REVIEW

File: api/src/controllers/cart.rs [hunk 1]
```diff
@@ -185,6 +194,14 @@ pub fn replace_cart(
     )?;
 
     cart.set_user_agent(request_info.user_agent.clone(), false, connection)?;
+    // Beware there could be multiple carts that meet this condition
+    for (ticket_type_id, remaining) in cart.ticket_types(connection)? {
+        println!("{:?} - {:?}", ticket_type_id, remaining);
```
Comment (philipr-za) [inline api/src/controllers/cart.rs]: Left some println's in here

File: api/src/controllers/ticket_types.rs [hunk 2]
```diff
@@ -254,6 +263,7 @@ pub fn update(
     )?;
 
     let data = data.into_inner();
+    //println!("{:?}", data);
```
Comment (philipr-za) [inline api/src/controllers/ticket_types.rs]: Remove

File: db/src/models/ticket_types.rs [hunk 3]
```diff
@@ -239,6 +304,93 @@ impl TicketType {
...
+        current_user_id: Option<Uuid>,
+        conn: &PgConnection,
+    ) -> Result<(), DatabaseError> {
+        if self.valid_available_ticket_count(conn)? > 0 {
+            return Ok(());
+        }
+
+        // Find child ticket types
+        for child in self.find_dependent_ticket_types(conn)? {
+            println!("child {:?}", child);
```
Comment (philipr-za) [inline db/src/models/ticket_types.rs]: Remove debug message

File: db/src/models/ticket_types.rs [hunk 4]
```diff
@@ -239,6 +304,93 @@ impl TicketType {
...
+                format!("Ticket sales have started for '{}'", self.name),
+                Tables::TicketTypes,
+                Some(self.id),
+                current_user_id,
+                Some(json!({"old_start_date": old_start_date, "new_start_date": self.start_date })),
+            )
+            .commit(conn)?;
+
+            let pricings = self.ticket_pricing(true, conn)?;
+            println!("pricings : {:?}", pricings);
```
Comment (philipr-za) [inline db/src/models/ticket_types.rs]: some more debug messages

File: db/src/models/enums.rs [hunk 5]
```diff
@@ -110,7 +110,16 @@ string_enum! { DomainEventTypes [
     TicketInstanceNullified,
     TicketInstancePurchased,
     TicketInstanceRedeemed,
-    TicketInstanceReleasedFromHold
+    TicketInstanceReleasedFromHold,
```
Comment (philipr-za) [inline db/src/models/enums.rs]: A question for my understanding. You are making quite a few DomainEvent types here but they don't seem to actually be triggering anything. Are you doing this in preparation for having them trigger stuff in the future?

File: db/src/models/ticket_pricing.rs [hunk 6]
```diff
@@ -186,30 +200,85 @@ impl TicketPricing {
             .to_db_error(ErrorCode::QueryError, "Could not load order_items")
     }
 
-    pub fn destroy(&self, conn: &PgConnection) -> Result<usize, DatabaseError> {
+    pub fn destroy(
+        &self,
+        current_user_id: Option<Uuid>,
```
Comment (philipr-za) [inline db/src/models/ticket_pricing.rs]: Is this part of a push to record who specifically did certain things in the system for audit?

File: db/src/models/enums.rs (hunk 5 above)
Comment (mikethetike) [reply db/src/models/enums.rs]: They also form an audit trail and at a later stage could trigger webhooks

File: db/src/models/ticket_pricing.rs (hunk 6 above)
Comment (mikethetike) [reply db/src/models/ticket_pricing.rs]: Yeah, for example: Who reduced the number of tickets

**Pull request merged**